File: Sem3/MAP/Laborator/Lab7/BasketballCompetition/BasketballCompetition/Repository/Memory/InMemoryRepository.cs
using System;
using System.Collections.Generic;
using BasketballCompetition.Domain;
using BasketballCompetition.Domain.Validators;

namespace BasketballCompetition.Repository.Memory
{
    public class InMemoryRepository<TId, TE> : IRepository<TId, TE> where TE : Entity<TId>
    {
        private readonly IValidator<TE> _validator;
        private readonly IDictionary<TId, TE> _entities;

        protected InMemoryRepository(IValidator<TE> validator)
        {
            _validator = validator;
            _entities = new Dictionary<TId, TE>();
        }

        public TE FindOne(TId id)
        {
            _entities.TryGetValue(id, out var rez);
            return rez;
        }

        public IEnumerable<TE> FindAll()
        {
            return _entities.Values;
        }

        public virtual TE Save(TE entity)
        {
            if (entity == null)
            {
                throw new Exception("Entity is null");
            }

            _validator.Validate(entity);
            
            if (_entities.ContainsKey(entity.Id))
            {
                throw new Exception("Entity is already in memory");
            }

            _entities[entity.Id] = entity;
            return default;
        }

        public TE Delete(TId id)
        {
            var rez = _entities[id];
            
            return _entities.Remove(id) ? rez : null;
        }

        public TE Update(TE entity)
        {
            if (entity == null)
            {
                throw new Exception("Entity is null");
            }
            
            _validator.Validate(entity);
            
            if (!_entities.ContainsKey(entity.Id)) return null;

            _entities[entity.Id] = entity;
            return entity;
        }
        
    }
}

File: Sem3/MAP/Laborator/Vacanta/src/main/java/repo/memory/InMemoryRepository.java
package repo.memory;

import domain.Entity;
import domain.validators.Validator;
import repo.RepoException;
import repo.Repository;

import java.util.*;

public class InMemoryRepository<ID, E extends Entity<ID>> implements Repository<ID, E> {
    private final Validator<E> validator;
    protected Map<ID, E> entities;


    public InMemoryRepository(Validator<E> validator) {
        this.validator = validator;
        entities = new HashMap<>();
    }


    @Override
    public Optional<E> findOne(ID id) {
        if (id == null) {
            throw new RepoException("Id is null!\n");
        }
        return Optional.ofNullable(entities.get(id));
    }

    @Override
    public List<E> findAll() {
        return (List<E>) entities.values();
    }

    @Override
    public Optional<E> save(E entity) {
        if (entity == null) {
            throw new RepoException("Entity is null!\n");
        }

        validator.validate(entity);

        if (entities.get(entity.getId()) != null) { // daca exista
            return Optional.of(entity);
        } else { // daca nu exista
            entities.put(entity.getId(), entity);
        }
        return Optional.empty();
    }

    @Override
    public Optional<E> delete(ID id) {
        if (id == null) {
            throw new RepoException("Id is null!\n");
        }

        return Optional.ofNullable(entities.remove(id));
    }

    @Override
    public Optional<E> update(E entity) {
        if (entity == null) {
            throw new RepoException("Entity is null!\n");
        }

        validator.validate(entity);

        if (entities.get(entity.getId()) != null) {
            entities.put(entity.getId(), entity);
            return Optional.empty();
        }

        return Optional.of(entity);
    }

    @Override
    public int size() {
        return entities.size();
    }
}
